ic-transport-types/src/lib.rs

```diff
@@ -122,9 +125,75 @@ pub enum QueryResponse {
     Replied {
         /// The reply from the canister.
         reply: ReplyResponse,
+
+        /// The list of node signatures.
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        signatures: Vec<NodeSignature>,
     },
     /// The request was rejected.
-    Rejected(RejectResponse),
+    Rejected {
+        /// The rejection from the canister.
+        #[serde(flatten)]
+        reject: RejectResponse,
+
+        /// The list of node signatures.
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        signatures: Vec<NodeSignature>,
+    },
+}
+
+impl QueryResponse {
+    /// Returns the signable form of the query response, as described in
+    /// [the spec](https://internetcomputer.org/docs/current/references/ic-interface-spec#http-query).
+    /// This is what is signed in the `signatures` fields.
+    pub fn signable(&self, request_id: RequestId, timestamp: u64) -> Vec<u8> {
+        #[derive(Serialize)]
+        #[serde(tag = "status", rename_all = "snake_case")]
+        enum QueryResponseSignable<'a> {
+            Replied {
+                reply: RequestId, // polyfill until hash_of_map is figured out
```

Comment:
All subnets including NNS has upgraded to support nested maps. This can be changed accrodingly.